sensor/kubernetes/listener/resources/networkpolicy.go

```diff
@@ -2,41 +2,75 @@ package resources
+	detector        detector.Detector
 }
 
-func newNetworkPolicyDispatcher(nps store.NetworkPolicyStore) *networkPolicyDispatcher {
+func newNetworkPolicyDispatcher(networkPolicyStore store.NetworkPolicyStore, deploymentStore *DeploymentStore, detector detector.Detector) *networkPolicyDispatcher {
 	return &networkPolicyDispatcher{
-		store: nps,
+		netpolStore:     networkPolicyStore,
+		deploymentStore: deploymentStore,
+		detector:        detector,
 	}
 }
 
-// ProcessEvent processes a network policy resource event, and returns the sensor events to generate.
-func (h *networkPolicyDispatcher) ProcessEvent(newObj, _ interface{}, action central.ResourceAction) []*central.SensorEvent {
-	np := newObj.(*networkingV1.NetworkPolicy)
-	netPolicy := networkPolicyConversion.KubernetesNetworkPolicyWrap{NetworkPolicy: np}.ToRoxNetworkPolicy()
+// ProcessEvent processes a network policy resource event and returns the sensor events to generate.
+func (h *networkPolicyDispatcher) ProcessEvent(obj, old interface{}, action central.ResourceAction) []*central.SensorEvent {
+	np := obj.(*networkingV1.NetworkPolicy)
 
-	switch action {
-	case central.ResourceAction_CREATE_RESOURCE, central.ResourceAction_UPDATE_RESOURCE:
-		h.store.Upsert(netPolicy)
-	case central.ResourceAction_REMOVE_RESOURCE:
-		h.store.Delete(netPolicy.GetId(), netPolicy.GetNamespace())
+	roxNetpol := networkPolicyConversion.KubernetesNetworkPolicyWrap{NetworkPolicy: np}.ToRoxNetworkPolicy()
+
+	if features.NetworkPolicySystemPolicy.Enabled() {
+		var roxOldNetpol *storage.NetworkPolicy
+		oldNp := old.(*networkingV1.NetworkPolicy)
+		if oldNp != nil {
+			roxOldNetpol = networkPolicyConversion.KubernetesNetworkPolicyWrap{NetworkPolicy: oldNp}.ToRoxNetworkPolicy()
+		}
+		sel := h.getSelector(roxNetpol, roxOldNetpol)
+		if action == central.ResourceAction_REMOVE_RESOURCE {
+			h.netpolStore.Delete(roxNetpol.GetId(), roxNetpol.GetNamespace())
+		} else {
+			h.netpolStore.Upsert(roxNetpol)
+		}
+
+		h.updateDeploymentsFromStore(roxNetpol, sel)
 	}
 
 	return []*central.SensorEvent{
 		{
 			Id:     string(np.UID),
 			Action: action,
 			Resource: &central.SensorEvent_NetworkPolicy{
-				NetworkPolicy: networkPolicyConversion.KubernetesNetworkPolicyWrap{NetworkPolicy: np}.ToRoxNetworkPolicy(),
+				NetworkPolicy: roxNetpol,
 			},
 		},
 	}
 }
+
+func (h *networkPolicyDispatcher) getSelector(np, oldNp *storage.NetworkPolicy) selector {
+	var sel selector
+	if oldNp != nil {
+		sel = MatcherOrEverything(oldNp.GetSpec().GetPodSelector().GetMatchLabels())
+	} else {
+		sel = labels.Nothing()
+	}
+	return or(sel, MatcherOrEverything(np.GetSpec().GetPodSelector().GetMatchLabels()))
+}
+
+func (h *networkPolicyDispatcher) updateDeploymentsFromStore(np *storage.NetworkPolicy, sel selector) {
+	for _, deploymentWrap := range h.deploymentStore.getMatchingDeployments(np.GetNamespace(), sel) {
+		h.detector.ProcessDeployment(deploymentWrap.GetDeployment(), central.ResourceAction_UPDATE_RESOURCE)
```

Comment:
I have just noticed that we have hardcoded action `central.ResourceAction_UPDATE_RESOURCE` here (no matter whether we upsert or delete). `ProcessDeployment` behaves differently for `REMOVE_RESOURCE` - are we sure that this value can be hardcoded?

Reply:
I think it has to be `update`, no? Because we're reacting to adding, updating, or removing a `NetworkPolicy`. So regardless of what action is, if it was associated with that deployment, we need to pretend that the deployment was updated in order to trigger the evaluation.

Reply:
I looked at `processDeploymentNoLock` and I think we are safe. It would be important if we would remove a deployment - then we would need to close violations for it. But as we only change the NetworkPolicies, then the only action making sense is UPDATE.